<a href="https://colab.research.google.com/github/StevenWestmoreland/DS-Unit-1-Sprint-2-Statistics/blob/master/DSPT6_Unit1_Sprint2_ProbabilityStatisticsAndInference_StevenWestmoreland.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [24]:
# import
import pandas as pd
from scipy.stats import ttest_ind

! wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data --no-check-certificate

--2020-04-09 01:42:45--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-04-09 01:42:46 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [25]:
column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

voters = pd.read_csv('house-votes-84.data', header=None, names=column_headers, na_values='?')
voters

party handicapped-infants water-project  ... crime duty-free south-africa
0    republican                   n             y  ...     y         n            y
1    republican                   n             y  ...     y         n          NaN
2      democrat                 NaN             y  ...     y         n            n
3      democrat                   n             y  ...     n         n            y
4      democrat                   y             y  ...     y         y            y
..          ...                 ...           ...  ...   ...       ...          ...
430  republican                   n             n  ...     y         n            y
431    democrat                   n             n  ...     n         n            y
432  republican                   n           NaN  ...     y         n            y
433  republican                   n             n  ...     y         n            y
434  republican                   n             y  ...     y       NaN            n

[435 rows x 17 columns]

In [28]:
# recode voters as numeric

voters = voters.replace({'y':1, 'n':0})
voters.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [29]:
# check corsstabs to find a comparison with more Dems voting Y than Rep

pd.crosstab(voters['party'], voters['budget'])

budget      0.0  1.0
party               
democrat     29  231
republican  142   22

In [30]:
# create republican and democrat dataframes 
rep = voters[voters['party']=='republican']
dem = voters[voters['party']=='democrat']

print(rep.shape)
print(dem.shape)

(168, 17)
(267, 17)


In [33]:
# check pvalue
ttest_ind(dem['budget'], rep['budget'], nan_policy='omit').pvalue

2.0703402795404463e-77

In regards to budget, Democrats voted yes more than Republicans with a pvalue of < 0.01 (actual result was approx. 2.07e-77)

In [34]:
# Now trying to find a comparison of Rep voting Y more than Dems

pd.crosstab(voters['party'], voters['education'])

education   0.0  1.0
party               
democrat    213   36
republican   20  135

In [40]:
# check pvalue

ttest_ind(dem['education'], rep['education'], nan_policy='omit').pvalue

1.8834203990450192e-64

Conversely, in regards to education Republicans voted yes more than Democrats with a pvalue of < 0.01 (actual result was approx. 1.88e-64). 

We've a 99.9% confidence that the votes for both Budget and Education are indicative of the population as a whole.

In [39]:
# looking for a comparison where the party seems independent of voting practices

pd.crosstab(voters['party'], voters['water-project'])

water-project  0.0  1.0
party                  
democrat       119  120
republican      73   75

In [42]:
# check pvalue

ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit').pvalue

0.9291556823993485

Because we've a pvalue of > 0.1 (approx. 0.93), we do not reject the Null Hypothesis in regards to Democrats and Republicans voting Yes on water-project.

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!